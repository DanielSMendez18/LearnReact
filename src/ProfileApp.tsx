import { useState } from "react";

interface Company {
    name: string,
    country: string
}

interface Skill {
    skillName: string,
    score: number
}

const ProfileApp = ({name, country}: Company) => {
    const [nameUser, setName] = useState<string>('');
    const [age, setAge] = useState<number>(18);
    const [hobbies, setHobbies] = useState<string[]>(['Reading','Gaming']);
    const [skill, setSkill] = useState<Skill>({skillName:'Front End', score: 0});

    const updateName = () => {
        setName('Daniel S.M')
    }

    const incrementAge = () => {
        setAge( age => age+1)
    }

    const addHobbie = (newHobbie: string) => {
        setHobbies(prevHobbies => [...prevHobbies, newHobbie])
    }

    const updateScoreSkill = (newScore: number) => {
        setSkill(prevSkill => ({...prevSkill, score: newScore}) )
    }

    const increaseScoreSkill = () => {
        const newScore = skill.score + 1;
        updateScoreSkill(newScore);
    }

    return(
        <>
            <h1>Compa√±ia: {name} - {country}</h1>
            <h2>Perfil del usuario</h2>
            <h3 aria-label="name-user" >Nombre: {nameUser}</h3>
            <h3 data-testid="test-value-year">Edad: {age}</h3>
            <button aria-label="update-name" onClick={updateName}>Actualizar nombre</button>
            <button data-testid="test-increment-year" onClick={incrementAge}>Aumentar edad</button>
            <br />
            <h3>Sus Hobies</h3>
            <ul>
                {hobbies.map((hobby, index) => (
                    <li key={index} >{hobby}</li>
                ))}
            </ul>
            <button aria-label="add-hobby" onClick={() => {addHobbie('Nuevo hobbie')}}>Agregar nuevo Hobbie</button>
            <br />
            <h3>Skill</h3>
            <h3 data-testid="test-value-skill">{skill.skillName} - Score: {skill.score}</h3>
            <button onClick={increaseScoreSkill} data-testid="test-increase-score">Incrementar Score</button>
        </>
    )

}

export default ProfileApp;