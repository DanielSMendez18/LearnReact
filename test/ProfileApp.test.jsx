import { fireEvent, render, screen } from '@testing-library/react'
import ProfileApp from '../src/ProfileApp'

describe('@ProfileApp', () => {
  const name = 'Globant'
  const country = 'ARG'

  test('#should to make match with the snapshop ', () => {
    //Arrange
    const { container } = render(<ProfileApp name={name} country={country} />)
    //Assert
    expect(container).toMatchSnapshot()
  })

  test('#should render the initial value to name in section CompaÃ±ia', () => {
    //Arrange
    render(<ProfileApp name={name} country={country} />)
    //Assert
    expect(screen.getByRole('heading', { level: 1 }).innerHTML).toContain(name)
  })

  test('#should update the name, after clicking "Actualizar nombre"', () => {
    //Arrange
    render(<ProfileApp name={name} country={country} />)
    //Act
    fireEvent.click(screen.getByRole('button', { name: 'update-name' }))
    const nameUpdated = screen.getByRole('heading', {
      name: 'name-user'
    }).innerHTML
    //Assert
    expect(nameUpdated).toContain('Daniel S.M')
  })

  test('#should increment the age value, after clicking "Aumentar Edad" ', () => {
    //Arrange
    render(<ProfileApp name={name} country={country} />)
    //Act
    fireEvent.click(screen.getByTestId('test-increment-year'))
    //Assert
    expect(screen.getByTestId('test-value-year').innerHTML).toContain('19')
  })

  test('#should add new hobby, after cliking "Agregar nuevo hobbie" button', () => {
    //Arrange
    render(<ProfileApp name={name} country={country} />)
    //Act
    const initialListHobbies = screen.getAllByRole('listitem')
    fireEvent.click(screen.getByRole('button', { name: 'add-hobby' }))
    const updatedListHobbies = screen.getAllByRole('listitem')
    //Assert
    expect(updatedListHobbies).toHaveLength(initialListHobbies.length + 1)
  })

  test('#should increment the score value, after clicking "Incrementar Score" ', () => {
    //Arrange
    render(<ProfileApp name={name} country={country} />)
    //Act
    fireEvent.click(screen.getByTestId('test-increase-score'))
    //Assert
    expect(screen.getByTestId('test-value-skill').innerHTML).toContain('1')
  })
})
