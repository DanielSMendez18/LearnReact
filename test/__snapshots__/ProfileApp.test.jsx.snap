// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@ProfileApp #should to make match with the snapshop  1`] = `
<div>
  <h1>
    Compa√±ia: 
    Globant
     - 
    ARG
  </h1>
  <h2>
    Perfil del usuario
  </h2>
  <h3
    aria-label="name-user"
  >
    Nombre: 
  </h3>
  <h3
    data-testid="test-value-year"
  >
    Edad: 
    18
  </h3>
  <button
    aria-label="update-name"
  >
    Actualizar nombre
  </button>
  <button
    data-testid="test-increment-year"
  >
    Aumentar edad
  </button>
  <br />
  <h3>
    Sus Hobies
  </h3>
  <ul>
    <li>
      Reading
    </li>
    <li>
      Gaming
    </li>
  </ul>
  <button
    aria-label="add-hobby"
  >
    Agregar nuevo Hobbie
  </button>
  <br />
  <h3>
    Skill
  </h3>
  <h3
    data-testid="test-value-skill"
  >
    Front End
     - Score: 
    0
  </h3>
  <button
    data-testid="test-increase-score"
  >
    Incrementar Score
  </button>
</div>
`;
